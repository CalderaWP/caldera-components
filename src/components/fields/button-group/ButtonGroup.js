import React from 'react';
import {optionShape} from '../propTypes';
import PropTypes from 'prop-types';
/**
 * Creates a button group field
 *
 * @param {Object} props
 * @returns {*}
 * @constructor
 */
export const ButtonGroup = (props) => {
	/**
	 * Dispatches new value(s) to parent
	 *
	 * @param selection
	 * @returns {*}
	 */
	function changeHandler(selection) {
		return props.onChange(selection);
	}

	/**
	 * Render a button group
	 */
	return (
		<div
			role="group"
		>
			{props.options.map(option => {
				return <button
					aria-label={option.ariaLabel ? option.ariaLabel : option.label}
					key={option.value}
					className={props.value === option.value ? 'selected' : 'not-selected' }
					onClick={() => {
						return changeHandler(option.value);
					}}
				>
					{option.hasOwnProperty('icon')&&
						<span className={option.icon}></span>
					}
					{!option.hasOwnProperty('icon') &&
						<React.Fragment>{option.label}</React.Fragment>
					}

				</button>;
			})}
		</div>
	);
};

/**
 * Prop definition for select fields
 *
 * @type {{}}
 */
ButtonGroup.propTypes = {
	onChange: PropTypes.func.isRequired,
	options: PropTypes.arrayOf(
		PropTypes.shape({
			...optionShape,
			icon:PropTypes.string,
			ariaLabel:PropTypes.string
		})
	),
	value: PropTypes.oneOfType([
		PropTypes.number,
		PropTypes.string
	])
};

/**
 * Default props for select fields
 *
 * @type {{options: Array}}
 */
ButtonGroup.defaultProps = {
	options: []
};