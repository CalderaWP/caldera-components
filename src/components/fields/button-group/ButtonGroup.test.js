import {ButtonGroup} from './ButtonGroup';
import renderer from 'react-test-renderer';
import React from 'react';
import {mount} from 'enzyme';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({adapter: new Adapter()});

describe('Button Group component', () => {
	function genericChangeHandler() {

	}

	describe( 'options', () =>{
		it('Creates with one option', () => {
			const component = renderer.create(
				<ButtonGroup
					onChange={genericChangeHandler}
					options={[
						{
							value: 'Tags',
							label: 'Tags'
						}
					]}
				/>
			);
			expect(component.toJSON()).toMatchSnapshot();
		});
		it('Renders two buttons', () => {
			let updateValue = null;
			const component = mount(
				<ButtonGroup
					onChange={(newValue) => {
						updateValue = newValue;
					}}
					options={[
						{
							value: 'tags',
							label: 'Tags'
						},
						{
							value: 'categories',
							label: 'Categories'
						}
					]}
				/>
			);
			expect(component.find('button')).toHaveLength(2);
		});

		it('Sends the right update value when clicked', () => {
			let updateValue = null;
			const component = mount(
				<ButtonGroup
					onChange={(newValue) => {
						updateValue = newValue;
					}}
					options={[
						{
							value: 'Tags',
							label: 'Tags'
						}
					]}
				/>
			);
			component.find('button').simulate('click');
			expect(updateValue).toBe('Tags');
		});

		it('Marks the value as selected if it should on initial load', () => {
			const component = mount(
				<ButtonGroup
					onChange={() => {

					}}
					options={[
						{
							value: 'a1',
							label: 'Tags'
						}
					]}
					value={'a1'}
				/>
			);

			expect(component.find('button').hasClass('selected')).toBe(true);
		});

		it('Sets aria-label to label  when ariaLabel not provided', () => {
			const component = mount(
				<ButtonGroup
					onChange={genericChangeHandler}
					options={[
						{
							value: 'a1',
							label: 'Tags',
						}
					]}
				/>
			);

			expect(component.find('button').prop('aria-label')).toBe('Tags');
		});

		it('Sets aria-label to explicit value when passed', () => {
			const component = mount(
				<ButtonGroup
					onChange={genericChangeHandler}
					options={[
						{
							value: 'a1',
							label: 'Tags',
							ariaLabel:'Select field values'
						}
					]}
				/>
			);

			expect(component.find('button').prop('aria-label')).toBe('Select field values');
		});
	});

	describe( 'icon buttons', ()=> {
		it('Creates with an icon option', () => {
			const component = renderer.create(
				<ButtonGroup
					onChange={genericChangeHandler}
					options={[
						{
							value: 'Tags',
							label: 'Tags',
							icon: 'fa-envelope'
						}
					]}
				/>
			);
			expect(component.toJSON()).toMatchSnapshot();
		});
	});


});