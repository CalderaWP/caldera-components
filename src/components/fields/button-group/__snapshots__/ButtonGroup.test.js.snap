// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Group component icon buttons Creates with an icon option 1`] = `
<div
  role="group"
>
  <button
    aria-label="Tags"
    className="not-selected"
    onClick={[Function]}
  >
    <span
      className="fa-envelope"
    />
  </button>
</div>
`;

exports[`Button Group component options Creates with one option 1`] = `
<div
  role="group"
>
  <button
    aria-label="Tags"
    className="not-selected"
    onClick={[Function]}
  >
    Tags
  </button>
</div>
`;
